### PR TITLE
new Runner, ProcessOutputInterface, DifferInterface and ResultInterface

### PR DESCRIPTION
**Pull request by keradus**
Introduce: new Runner, ProcessOutputInterface, DifferInterface and ResultInterface
Alternative for #1618 

```bash
D:\GIT\fork\PHP-CS-Fixer (master)
time php php-cs-fixer fix src
Loaded config from "D:\GIT\fork\PHP-CS-Fixer\.php_cs.dist"
Fixed all files in 42.924 seconds, 8.000 MB memory used
```

```bash
D:\GIT\fork\PHP-CS-Fixer (2.x-async_linting)
time php php-cs-fixer fix src
Loaded config from "D:\GIT\fork\PHP-CS-Fixer\.php_cs.dist"
Fixed all files in 27.270 seconds, 8.000 MB memory used
```

### CODE REVIEW

**Comment by keradus** [comment]
This is just a proof of concept, if we agreed to go that way it will need more work like filling the docs and fix tests

**Comment by gharlan** [comment]
At the moment the linting process is started for all files, regardless if they are cached or not..

**Comment by keradus** [comment]
:(

**Comment by GrahamCampbell** [comment]
> At the moment the linting process is started for all files, regardless if they are cached or not..

Yeh, there's no way this is allowed. ;(

File: src/Linter/ProcessResult.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+namespace Symfony\CS\Linter;
+
+use Symfony\Component\Process\Process;
+
+/**
+ * TODO.
+ *
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ */
+final class ProcessResult
```
Comment (stof) [inline src/Linter/ProcessResult.php]: missing interface implementation

File: src/Linter/NullResult.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+namespace Symfony\CS\Linter;
+
+/**
+ * TODO.
+ *
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ */
+final class NullResult
```
Comment (stof) [inline src/Linter/NullResult.php]: missing interface implementation

Comment (stof) [inline src/Linter/NullResult.php]: missing interface implementation

**Comment by keradus** [comment]
Yeah, as I sad it is a POC, is it worth to go further ?

**Comment by gharlan** [comment]
> is it worth to go further ?

:+1:

File: src/FileLintingIterator.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+namespace Symfony\CS;
+
+use Symfony\CS\Linter\LinterInterface;
+
+/**
+ * TODO
+ *
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ */
+final class FileLintingIterator extends \CachingIterator
```
Comment (gharlan) [inline src/FileLintingIterator.php]: is it intended intended that there are two nested CashingIterators?

Comment (keradus) [inline src/FileLintingIterator.php]: Nope, this one is enought. good catch

**Comment by keradus** [comment]
> At the moment the linting process is started for all files, regardless if they are cached or not..

For the current master we have two *public* methods in main `Fixer` class:
- `fix` - for fixing everything
- `fixFile` for fixing single file
For now, the `fixFile` is taking care of cacheing and linting and `fix` simply iterate over all files and calls `fixFile` for each of them.
We want to lint one file ahead, so it should be done in `fix` method. But also we don't want to lint if cache is valid, so we need to move cache checking to `fix` as well.

For that I think we should encapsulate `fixFile` method and only expose interface to fix things in bulk process (one could always simply pass collection of files to be fixed with whole one element).
I have started to move logic from `fixFile` to `fix` and fixed the issue with cache.

It just show the direction, the refactoring of `fix` vs `fixFile` is not done yet.

What do you think ?

File: src/Fixer.php [hunk 4]
```diff
@@ -166,22 +200,12 @@ public function fixFile(\SplFileInfo $file, array $fixers, $dryRun, $diff, FileC
...
-
-            return;
-        }
-
         try {
-            $this->linter->lintFile($file->getRealPath());
+            if (null !== $this->currentFileLintingResult) {
+                $this->currentFileLintingResult->check();
+            } else {
+                $this->linter->lintFile($file->getRealPath())->check();
```
Comment (keradus) [inline src/Fixer.php]: this line could be removed when `fixFile` will become private

**Comment by keradus** [comment]
Any thoughts?

**Comment by gharlan** [comment]
I think, the direction is good! :+1:  ;)

**Comment by keradus** [comment]
Then `Fixer` class needs a big rewrite...

**Comment by keradus** [comment]
I would made it more state aware... passing everything to every method as parameter is pointless...

**Comment by keradus** [comment]
Runner of fixers has been rewritten from scratch.
A bit more work needed for linting in async way

**Comment by keradus** [comment]
(Please ignore 5.3 for now)

**Comment by keradus** [comment]
Sorry for taking so long.
I have 5 branches with different approach for this problem. This one works the best.

I updated first post.

**Comment by keradus** [comment]
@gharlan please, could you take a look?

File: src/Runner/Runner.php [hunk 5]
```diff
@@ -10,100 +10,110 @@
...
      * @var ErrorsManager
      */
-    protected $errorsManager;
+    private $errorsManager;
 
     /**
-     * Linter instance.
-     *
-     * @var LinterInterface
+     * @var type FileCacheManager
```
Comment (gharlan) [inline src/Runner/Runner.php]: type?

File: src/Runner/FileLintingIterator.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            $this->nextResult = $this->handleItem($this->getInnerIterator()->current());
+        }
+    }
+
+    private function handleItem(\SplFileInfo $file)
+    {
+        return $this->linter->lintFile($file->getRealPath());
+    }
+
+    private function getFileRelativePathname(\SplFileInfo $file)
```
Comment (gharlan) [inline src/Runner/FileLintingIterator.php]: method is not used

File: src/Runner/FileFilterIterator.php [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+use PhpCsFixer\FileCacheManager;
+use PhpCsFixer\FixerFileProcessedEvent;
+use Symfony\Component\EventDispatcher\Event;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+/**
+ * @internal
+ */
+final class FileFilterIterator extends \FilterIterator
```
Comment (gharlan) [inline src/Runner/FileFilterIterator.php]: author?

**Comment by gharlan** [comment]
looks very well, thanks for the work!

File: src/Runner/FileLintingIterator.php (hunk 6 above)
Comment (keradus) [inline src/Runner/FileLintingIterator.php]: I think it's residue from another approach I have tried, nice catch

**Comment by keradus** [comment]
Thank you @gharlan , I applied your comments ;)

File: src/Differ/SebastianBergmannDiffer.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        $this->differ = new Differ();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function diff($old, $new)
+    {
+        return implode(
+            PHP_EOL,
```
Comment (GrahamCampbell) [inline src/Differ/SebastianBergmannDiffer.php]: Might be an idea to have this configurable via an optional constructor param, so a windows system can output unix style line endings if desired.

Comment (keradus) [inline src/Differ/SebastianBergmannDiffer.php]: This PR is already big, what you are referring to is not changed - it's also current, master implementation. It's outside the scope of this PR.
My preferred option would be to not have it at all ;)

**Pull request merged**